### PR TITLE
removeObjects API - Removes all the objects residing under a bucket (API)

### PR DESCRIPTION
**Pull request by Praveenrajmani**
Function **removeObjects(bucketName, objectsCh)**

objectsCh - Channel of objects to be removed ( used [this](https://github.com/NodeGuy/channel) for channels )

Behaves exactly like go sdk's [removeObjects](https://github.com/minio/minio-go/blob/master/docs/API.md#RemoveObjects) , returns a stream of errors 

Fixes: #695

### CODE REVIEW

**Comment by krishnasrinivas** [comment]
Since channel style programming is not native to JS, don't do it this way.

You can have a simpler API:
```
removeObjects(bucketName, [object1, object2, object3, ...], cbk)
```

The other option is:
```
removeObjects(bucketName, stream, cbk)
```
but it is harder for developers to consume this API (channels in Go are easy, streams in JS are hard)

Hence you can do ```removeObjects(bucketName, [object1, object2, object3, ...], cbk)```

**Comment by Praveenrajmani** [comment]
Hi @krishnasrinivas  , yes got it 
Made the changes , So now the function reads from the list of objects (the first option)

**Comment by builtbyproxy** [comment]
Thanks for all your work @Praveenrajmani 
Im excited for this!

File: examples/remove-objects.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+var s3Client = new Minio.Client({
+  endPoint: 's3.amazonaws.com',
+  port: 9000,
+  secure: false,
+  accessKey: 'YOUR-ACCESSKEYID',
+  secretKey: 'YOUR-SECRETACCESSKEY'
+});
+
+var objectsList = []
```
Comment (kannappanr) [inline examples/remove-objects.js]: Maybe add a sample list of objects instead of leaving it empty.

File: src/main/minio.js [hunk 2]
```diff
@@ -1421,6 +1421,74 @@ export class Client {
...
+
+
+  generateRemoveMultiObjectsRequest(objects) {
+    var deleteObjects={"Delete":[{"Quiet": true}]}
+
+    objects.forEach(function(value){
+      deleteObjects["Delete"].push({"Object": [{"Key": value}]})
+    })
+
+    var xmlBytes = Xml(deleteObjects)
```
Comment (kannappanr) [inline src/main/minio.js]: you can do `return Xml(deleteObjects)` instead

**Comment by krishnasrinivas** [comment]
@Praveenrajmani javascript is a tricky language. The code for removeObjects is incorrect because of the way async behavior of JS. What will happen is `cb(null)` always gets called, if there is an error `return cb(e)` is also called.

I haven't tested the following code, but it should be written like this. Can you check with Kanakaraj on the explanation on the following code? he will be able to explain it better in person on the code flow.

There are some additional changes needed to parse the success response XML, I will tell you about it later which can be done once this is merged.

```
  removeObjects(bucketName, objectsList, cb) {
    if (!isValidBucketName(bucketName)) {
      throw new errors.InvalidBucketNameError('Invalid bucket name: ' + bucketName)
    }
    if (typeof objectsList !== "object") {
      throw new errors.InvalidArgumentError('objectsList should be a list')
    }
    if (!isFunction(cb)) {
      throw new TypeError('callback should be of type "function"')
    }
    const maxEntries = 1000
    const query = 'delete'
    const method = 'POST'

    let result = objectsList.reduce((result, entry) => {
      result.list.push(entry)
      if (result.list.length === maxEntries) {
        result.listOfList.push(result.list)
        result.list = []
      }
      return result
    }, {listOfList: [], list:[]})

    async.eachSeries(result.listOfList, (list, cb) => {
      var deleteObjects={"Delete":[{"Quiet": true}]}
      
      list.forEach(function(value){
        deleteObjects["Delete"].push({"Object": [{"Key": value}]})
      })

      let payload = Xml(deleteObjects)
      this.makeRequest({ method, bucketName, query}, payload, 200, '', false, (e) => {
        if (e) return cb(e)
        cb(null)
      })
    }, cb)
  }
```

**Comment by kannappanr** [comment]
@Praveenrajmani Can you also fix the git commit message

**Comment by Praveenrajmani** [comment]
Thanks for the explanation @krishnasrinivas @kannappanr  . Will do the changes and update

**Comment by Praveenrajmani** [comment]
@krishnasrinivas @kannappanr  made the changes

File: src/main/minio.js [hunk 3]
```diff
@@ -1421,6 +1421,61 @@ export class Client {
     this.makeRequest({method, bucketName, objectName}, '', 204, '', false, cb)
   }
 
+  // Remove all the objects residing in the objectsList.
+  //
+  // __Arguments__
+  // * `bucketName` _string_: name of the bucket
+  // * `objectsCh` _string_: channel of objects
```
Comment (kannappanr) [inline src/main/minio.js]: Typo, rename `objectCh` to `objectsList`

File: src/main/minio.js [hunk 4]
```diff
@@ -1421,6 +1421,61 @@ export class Client {
...
+  removeObjects(bucketName, objectsList, cb) {
+    if (!isValidBucketName(bucketName)) {
+      throw new errors.InvalidBucketNameError('Invalid bucket name: ' + bucketName)
+    }
+    if (typeof objectsList !== "object") {
```
Comment (kannappanr) [inline src/main/minio.js]: Can this be more specifically checked for an Arry/List?

File: src/main/minio.js (hunk 3 above)
Comment (Praveenrajmani) [reply src/main/minio.js]: *oops ! will change it

File: src/main/minio.js (hunk 4 above)
Comment (Praveenrajmani) [reply src/main/minio.js]: @kannappanr  Done.

File: docs/API.md [hunk 5]
```diff
@@ -616,6 +617,53 @@ minioClient.removeObject('mybucket', 'photo.jpg', function(err) {
...
+
+Remove all objects in the objectsList.
+
+__Parameters__
+
+
+| Param | Type | Description |
+| ---- | ---- | ---- |
+| `bucketName` | _string_ | Name of the bucket. |
+| `objectsList`  | _object_  |  list of objects in the bucket.  |
```
Comment (kannappanr) [inline docs/API.md]: This can be changed to `list of objects in the bucket to be removed`

**Comment by kannappanr** [comment]
@Praveenrajmani Everything looks good except for the minor document comment.

**Comment by builtbyproxy** [comment]
Woo! Good work @Praveenrajmani

File: docs/API.md (hunk 5 above)
Comment (Praveenrajmani) [reply docs/API.md]: changed it @kannappanr

**Comment by builtbyproxy** [comment]
Awesome work team! Thankyou very much all @kannappanr @krishnasrinivas @Praveenrajmani

**Pull request merged**